### PR TITLE
[Tosa] support reduce_sum/prod/max/min floating point constant fold

### PR DESCRIPTION
**Pull request by ge28boj**
https://github.com/Xilinx/llvm-project/pull/180 closed. Completely reverting the format is much more easier in this way.

### CODE REVIEW

File: mlir/test/Dialect/Tosa/constant-op-fold.mlir [hunk 1]
```diff
@@ -656,6 +656,16 @@ func.func @reverse_length_one(%arg0 : tensor<10x1xi32>) -> (tensor<10x1xi32>, te
     return %0 : tensor<1x3xi32>
   }
 
+// -----
+
+  func.func @reduce_sum_constant() -> tensor<1x3xf32> {
```
Comment (mgehre-amd) [inline mlir/test/Dialect/Tosa/constant-op-fold.mlir:661]: Can you please add tests for some other float types like bf16 and f16?

File: mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td [hunk 2]
```diff
@@ -1409,9 +1417,10 @@ def Tosa_ReduceMaxOp : Tosa_InferTensorTypeOp<"reduce_max"> {
     /// Method used by InferTypeOpInterface.
     static bool isCompatibleReturnTypes(TypeRange l, TypeRange r);
 
-    /// Return the max of the two integer operands
-    static inline APInt calcOneElement(APInt leftOperand, APInt rightOperand) {
-      const llvm::APInt subtractRes = leftOperand - rightOperand;
+   /// Return the max of the two operands
+    template<typename T>
+    static std::optional<T> calcOneElement(T leftOperand, T rightOperand) {
```
Comment (mgehre-amd) [inline mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td:1422]: Does this correctly implement the behavior for NaNs as specified in `apply_max_s` on https://www.mlplatform.org/tosa/tosa_spec.html#_arithmetic_helpers?
Same question for reduce_min

Comment (ge28boj) [reply mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td:1422]: good point! I add the conditional check for NaNs

**Review by mgehre-amd — APPROVED** [review]

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 3]
```diff
@@ -1716,29 +1735,35 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
...
+      llvm::SmallVector<APInt> newReducedTensor(newNumOfElements);
+      resultDenseAttr =
+          this->compute<APInt>(newReducedTensor, denseElementsAttr, oldShape,
+                               reductionAxis, rankedTensorType);
+    } else if (llvm::isa<FloatType>(shapedOldElementsValues.getElementType())) {
+      llvm::SmallVector<APFloat> newReducedTensor(newNumOfElements,
+                                                  APFloat(0.0));
+      resultDenseAttr =
+          this->compute<APFloat>(newReducedTensor, denseElementsAttr, oldShape,
+                                 reductionAxis, rankedTensorType);
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1763]: I don't know if it will happen of element type not be Integer or Float, but can we have an else statement here just to be safe? something like:
```suggestion
                                 reductionAxis, rankedTensorType);
                          } else {
                        return rewriter.notifyMatchFailure("unsupported element type");
                          
                          }
```

File: mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td [hunk 4]
```diff
@@ -1341,9 +1341,13 @@ def Tosa_ReduceAllOp : Tosa_InferTensorTypeOp<"reduce_all"> {
...
-    /// Return the AND result between two integer operands
-    static inline APInt calcOneElement(APInt leftOperand, APInt rightOperand) {
-      return leftOperand & rightOperand;
+    /// Return the AND result between two operands
+    template<typename T>
+    static std::optional<T> calcOneElement(T leftOperand, T rightOperand) {
+      if constexpr (std::is_same_v<T, APInt>) {
+        return leftOperand & rightOperand;
+      }
+      return std::nullopt;
```
Comment (ttjost) [inline mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td:1350]: What would happen for FP types in this case?

**Comment by ge28boj** [comment]
As suggested by @ttjost, it would be cleaner to get rid of `std::optional<T>` for `calcOneElement`.

**Review by mgehre-amd — APPROVED** [review]

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 5]
```diff
@@ -1716,29 +1734,42 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
     const auto shapedOldElementsValues =
         denseElementsAttr.getType().cast<ShapedType>();
 
-    if (!llvm::isa<IntegerType>(shapedOldElementsValues.getElementType()))
-      return rewriter.notifyMatchFailure(
-          op, "reduce input currently supported with integer type");
+    if constexpr (std::is_same_v<OperationType, ReduceAllOp> ||
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1737]: How is this check related to `FPSupport`? Could/should it be combined?

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 6]
```diff
@@ -1665,25 +1665,25 @@ llvm::APInt calculateReducedValue(const mlir::ElementsAttr &oldTensorAttr,
...
-    reducedValue =
-        OperationType::calcOneElement(reducedValue, oldTensor[index]);
+    reducedValue = OperationType::template calcOneElement<ElementType>(
+            reducedValue, oldTensor[index]);
   }
   return reducedValue;
 }
 
-template <typename OperationType>
+template <typename OperationType, bool FPSupport = true>
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1686]: maybe?
```suggestion
template <typename OperationType, bool hasFPSupport = true>
```

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 7]
```diff
@@ -1716,29 +1734,43 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
     const auto shapedOldElementsValues =
         denseElementsAttr.getType().cast<ShapedType>();
 
-    if (!llvm::isa<IntegerType>(shapedOldElementsValues.getElementType()))
-      return rewriter.notifyMatchFailure(
-          op, "reduce input currently supported with integer type");
+    if constexpr (FPSupport == false &&
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1737]: do you really need it here?

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 8]
```diff
@@ -1716,29 +1734,43 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
...
-      newReducedTensor[reductionIndex] = calculateReducedValue<OperationType>(
-          denseElementsAttr, oldShape, reductionAxis, reductionIndex);
+    DenseElementsAttr resultDenseAttr;
+    if (llvm::isa<IntegerType>(shapedOldElementsValues.getElementType())) {
+      llvm::SmallVector<APInt> newReducedTensor(newNumOfElements);
+      resultDenseAttr =
+          this->compute<APInt>(newReducedTensor, denseElementsAttr, oldShape,
+                               reductionAxis, rankedTensorType);
+    } else if (llvm::isa<FloatType>(shapedOldElementsValues.getElementType())) {
+      if constexpr(FPSupport == true) {
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1759]:
```suggestion
      if constexpr(hasFPSupport) {
```

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 9]
```diff
@@ -1716,29 +1734,43 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
...
+        llvm::SmallVector<APFloat> newReducedTensor(newNumOfElements,
+                                                    APFloat(0.0));
+        resultDenseAttr =
+            this->compute<APFloat>(newReducedTensor, denseElementsAttr, oldShape,
+                                  reductionAxis, rankedTensorType);  
+      } else if constexpr (FPSupport == false) {
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1765]:
```suggestion
      } else {
```

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp [hunk 10]
```diff
@@ -1716,29 +1734,43 @@ struct ReduceConstantOptimization : public OpRewritePattern<OperationType> {
...
+        return rewriter.notifyMatchFailure(
+            op, "FPSupport is false but element type is floating point");
```
Comment (ttjost) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1767]:
```suggestion
            op, "no support for floating point type.");
```

**Review by ttjost — APPROVED** [review]
I made some more comments.
@mgehre-amd , I asked Liang-Ta to change `calcOneElement` so that it doesn't return optional because it would never truly return it (we'd bail before). Could you check it once more? thanks

**Comment by ttjost** [comment]
> I made some more comments. @mgehre-amd , I asked Liang-Ta to change `calcOneElement` so that it doesn't return optional because it would never truly return it (we'd bail before). Could you check it once more? thanks

Nevermind, I just saw your comment.

File: mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp (hunk 7 above)
Comment (ttjost) [reply mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1737]: perhaps we can only keep `hasFPSupport` and remove the `is_same_v` part?

Comment (ge28boj) [reply mlir/lib/Dialect/Tosa/Transforms/TosaFolders.cpp:1737]: I think so, without 'FPSupport == false', it means the statement "`reduce_any` and `reduce_all` don't support floating point" is the source of truth. But actually the spec doesn't forbid input element type of `reduce_any` and `reduce_all` to be floating point, nor does the verifier check on that. We 'think' it makes no sense to have floating point as input element type because we use logical operators for APFloat as implementation.
